server/app/middlewares/websocket_backend.rb

```diff
@@ -36,10 +38,29 @@ def initialize(app)
     watchdog
   end
 
+  # @return [String]
+  def our_version
+    major, minor, patch, extension = Server::VERSION.split('.')
+    version = "#{major}.#{minor}.0"
+    version << ".#{extension}" if extension
+    version
+  end
+
+  # @param [String] agent_version
+  # @return [Boolean]
+  def valid_agent_version?(agent_version)
+    Gem::Dependency.new('', "~> #{self.our_version}").match?('', agent_version)
+  end
+
   def call(env)
     if Faye::WebSocket.websocket?(env)
       req = Rack::Request.new(env)
-      ws = Faye::WebSocket.new(env)
+      ws = Faye::WebSocket.new(env, nil,
+        headers: {
+          'Kontena-Version' => Server::VERSION,
```

Comment:
Considering using the websocket request headers to replace the `/agent/master_info` RPC, by having the request handler lookup and validate the node etc before actually opening the websocket.

The agent would use this to update its `/etc/kontena.env`?